File: src/c_cpp_src/OPHD/UI/NotificationWindow.cpp
#include "NotificationWindow.h"

#include "../Cache.h"
#include "../Constants.h"

#include <NAS2D/Utility.h>
#include <NAS2D/Renderer/Renderer.h>


using namespace NAS2D;


NotificationWindow::NotificationWindow():
	mIcons{imageCache.load("ui/icons.png")}
{
	size({300, 220});

	add(btnOkay, {245, 195});
	btnOkay.size({50, 20});

	add(btnTakeMeThere, {10, 195});
	btnTakeMeThere.size({125, 20});
	btnTakeMeThere.hide();

	add(mMessageArea, {5, 65});
	mMessageArea.size({size().x - 10, 125});
	mMessageArea.font(constants::FONT_PRIMARY, constants::FontPrimaryNormal);
}


void NotificationWindow::notification(const NotificationArea::Notification& notification)
{
	mNotification = notification;
	title(StringFromNotificationType(mNotification.type));
	mMessageArea.text(mNotification.message);

	title(mNotification.brief);

	mTakeMeThereVisible = mNotification.position != Point<int>{-1, -1}; //\fixme magic value
}


void NotificationWindow::btnOkayClicked()
{
	hide();
}


void NotificationWindow::btnTakeMeThereClicked()
{
	mTakeMeThereClicked(mNotification.position, mNotification.depth);
	hide();
}


void NotificationWindow::update()
{
	if (!visible()) { return; }

	Window::update();

	btnTakeMeThere.visible(mTakeMeThereVisible); // bit of a hack

	auto& renderer = Utility<Renderer>::get();

	Point<float> iconLocation = position() + Vector{10, 30};

	renderer.drawSubImage(mIcons, iconLocation, {128, 64, 32, 32}, ColorFromNotification(mNotification.type));
	renderer.drawSubImage(mIcons, iconLocation, IconRectFromNotificationType(mNotification.type), Color::Normal);
}

File: outpost_rs/src/OPHD/UI/NotificationWindow.rs
#pragma once

#include "Core/Button.h"
#include "Core/TextArea.h"
#include "Core/Window.h"

#include "NotificationArea.h"

#include <NAS2D/Signal/Signal.h>

class NotificationWindow : public Window
{

	using Signal = NAS2D::Signal<NAS2D::Point<int>, int>;


	NotificationWindow();

	void notification(const NotificationArea::Notification&);

	Signal& takeMeThere() { return mTakeMeThereClicked; }

	void update() override;


	void btnOkayClicked();
	void btnTakeMeThereClicked();

	const NAS2D::Image& mIcons;

	NotificationArea::Notification mNotification;
	Button btnOkay{"Okay", {this, &NotificationWindow::btnOkayClicked}};
	Button btnTakeMeThere{"Take Me There", {this, &NotificationWindow::btnTakeMeThereClicked}};
	TextArea mMessageArea;
	bool mTakeMeThereVisible{false};

	Signal mTakeMeThereClicked;
};
